[wasm] Modify workload to pick threading runtime packs

steveisok opened the pull request:
This change adds the 2 wasm threading runtime packs to the wasm workload. In order for a threading runtime pack to be chosen, WorkloadManifest.targets is also modified to override the runtime pack name when the following props are set:

WasmEnableThreads - full threading support and will load `Microsoft.NETCore.App.Runtime.multithread.Mono.browser-wasm`

WasmEnablePerfTrace - runtime only threading support and will load `Microsoft.NETCore.App.Runtime.perftrace.Mono.browser-wasm`

Contributes to #72872

Commit:
[wasm] Modify workload to pick threading runtime packs

This change adds the 2 wasm threading runtime packs to the wasm workload. In order for a threading runtime pack to be chosen, WorkloadManifest.targets is also modified to override the runtime pack name when the following props are set:

WasmEnableThreads - full threading support and will load Microsoft.NETCore.App.Runtime.multithread.Mono.browser-wasm

WasmEnablePerfTrace - runtime only threading support and will load Microsoft.NETCore.App.Runtime.perftrace.Mono.browser-wasm

lewing commented:
Need to make sure WBT knows about the packs

lewing reviewed (commented):
A couple of questions but should be looks good otherwise

steveisok commented on src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets.in:
```diff
@@ -93,6 +93,9 @@
       <KnownRuntimePack Update="@(KnownRuntimePack)">
         <LatestRuntimeFrameworkVersion Condition="'%(KnownRuntimePack.TargetFramework)' == '${NetCoreAppCurrent}' and '%(KnownRuntimePack.RuntimePackLabels)' == 'Mono'">**FromWorkload**</LatestRuntimeFrameworkVersion>
+        <!-- Overrides for wasm threading support -->
+        <RuntimePackNamePatterns Condition="'$(RuntimeIdentifier)' == 'browser-wasm' and '$(WasmEnableThreading)' == 'true'">Microsoft.NETCore.App.Runtime.Mono.multithread.**RID**</RuntimePackNamePatterns>
+        <RuntimePackNamePatterns Condition="'$(RuntimeIdentifier)' == 'browser-wasm' and '$(WasmEnablePerfTrace)' == 'true'">Microsoft.NETCore.App.Runtime.Mono.perftrace.**RID**</RuntimePackNamePatterns>
```
Perftrace is single threaded userspace, multi-threaded runtime.  Do you think we should error if both are true?

Commit:
Add error condition when both WasmEnableThreading and WasmEnablePerfTrace are true

radical commented:
Looks good. I can add a test in WBT.

radical commented:
`Workload installation failed: microsoft.netcore.app.runtime.mono.multithread.browser-wasm::7.0.0-ci is not found in NuGet feeds`. I don't see the this nuget being generated in the [build](https://dev.azure.com/dnceng/public/_build/results?buildId=1890352&view=logs&jobId=6f051c1a-c3e7-5cde-411e-64125624a208&j=6f051c1a-c3e7-5cde-411e-64125624a208&t=b731ef43-7f67-5a47-0aca-433fa8d3212b).

radical commented:
> To get this to work, I think WasmBuildTests is going to need to wait for the threading builds to complete and then import its runtime packs.

Ok, I will add that.

radical committed:
[wasm] Set default value for MonoWasmBuildVariant, for runtime pack nuget

radical committed:
[wasm] Wasm.Build.Tests: add support for multiple runtime packs

- Runtime pack nugets other than the one that gets built might not be
available. In that case, we copy the only one available as the missing
ones.
    - This is useful for local builds, allowing running Wasm.Build.Tests
      without having to build all 3 runtimes.
    - This is enabled for CI also, in this commit. But once CI gets the
      ability to fetch the other runtime packs, this will change.

radical committed:
Merge remote-tracking branch 'origin/main' into resolve-wasm-mt-packs

radical commented:
I have pushed some changes that will unblock this PR. Essentially, we copy the single runtime pack nuget to the other nuget filenames, and should allow running Wasm.Build.Tests . Once I have the PR, to get the other correct runtime packs on CI, completed, I will change this behavior to be applicable only for local builds.

It will emit this when it copies:
```
        ********************

        Note: Could not find the expected three runtime packs in /workspaces/runtime/artifacts/packages/Release/Shipping/. Found Microsoft.NETCore.App.Runtime.Mono.browser-wasm.7.0.0-dev, Microsoft.NETCore.App.Runtime.Mono.multith
read.browser-wasm.7.0.0-dev, Microsoft.NETCore.App.Runtime.Mono.perftrace.browser-wasm.7.0.0-dev .
              To support local builds, /workspaces/runtime/artifacts/packages/Release/Shipping/Microsoft.NETCore.App.Runtime.Mono.browser-wasm.7.0.0-dev.nupkg was copied as the missing nugets.

        *******************
```

radical committed:
[wasm] Wasm.Build.Tests: build runtime pack nugets with different names

- The earlier approach of simply making copies of the existing runtime
pack nuget with different names doesn't work, and `dotnet workload
install` rejects it.

```
  Installing pack Microsoft.NETCore.App.Runtime.Mono.multithread.browser-wasm version 7.0.0-ci...
  Workload installation failed. Rolling back installed packs...
```

Instead, now we build the missing nugets from the project with different
values for `$(MonoWasmBuildVariant)`.

- this handles local builds, and incremental builds also
- To skip building the missing nugets, for example, when you have all of
them available, then set `WasmSkipMissingRuntimePackBuild=true`.

radical committed:
fix comment

radical commented:
/azp run runtime-wasm

radical commented:
The coreclr failure seems to be https://github.com/dotnet/runtime/issues/68289 and unrelated to this PR.
The WBT failures are known (https://github.com/dotnet/runtime/issues/72880), and unrelated.
